### PR TITLE
remove chat template config in turbomind engine

### PR DESCRIPTION
**Pull request by irexyc**
## Motivation

Align to torch engine, remove chat_template_mdoel in turbomind engine.

**BC-breaking**
When using turbomind engine directly, user needs to explicitly specify EngineGenerationConfig and set stop_words

### CODE REVIEW

File: lmdeploy/serve/async_engine.py [hunk 1]
```diff
@@ -235,10 +235,7 @@ def _build_turbomind(
             backend_config.session_len = self.chat_template.session_len
         from lmdeploy import turbomind as tm
         self.engine = tm.TurboMind.from_pretrained(
-            model_path,
-            engine_config=backend_config,
-            chat_template_config=chat_template_config,
```
Comment (AllentDan) [inline lmdeploy/serve/async_engine.py:240]: may remove the `chat_template_config` argument of `_build_turbomind` function as well.

**Review by AllentDan — APPROVED** [review]
Tesed with
1. lmdeploy chat turbomind internlm2-chat-7b w/wo --chat-template
2. pipeline of internlm2-chat-7b w/wo --chat-template
3. api_server with chat-template

File: lmdeploy/serve/async_engine.py (hunk 1 above)
Comment (AllentDan) [reply lmdeploy/serve/async_engine.py:240]: may remove the argument when calling the `__build_turbomind`

File: lmdeploy/turbomind/chat.py [hunk 2]
```diff
@@ -20,24 +22,25 @@ def input_prompt(model_name):
...
 def main(model_path: str,
          model_name: str = None,
          session_id: int = 1,
+         top_k: float = 40,
+         top_p: float = 0.8,
+         temperature: float = 0.8,
+         repetition_penalty: float = 1.0,
          cap: str = 'chat',
          tp: int = 1,
+         max_batch_size: int = 128,
```
Comment (lvhan028) [inline lmdeploy/turbomind/chat.py:34]: 这里 max_batch_size 改成 1 吧

File: lmdeploy/turbomind/chat.py [hunk 3]
```diff
@@ -46,100 +49,115 @@ def main(model_path: str,
...
-        for k, v in kwargs.items():
-            if hasattr(chat_template_cfg, k):
-                setattr(chat_template_cfg, k, v)
-            else:
-                new_kwargs[k] = v
-        kwargs = new_kwargs
+        chat_template (str): user defined chat template
+        kwargs (dict): unused args
+    """ # noqa: E 501
+    print('unused kwargs', kwargs, sep='')
```
Comment (lvhan028) [inline lmdeploy/turbomind/chat.py:70]: 有必要打印么？

File: lmdeploy/turbomind/chat.py (hunk 2 above)
Comment (lvhan028) [inline lmdeploy/turbomind/chat.py:34]: max_batch_size 感觉不用作为参数

File: lmdeploy/turbomind/chat.py [hunk 4]
```diff
@@ -46,100 +49,115 @@ def main(model_path: str,
...
-        if hasattr(engine_cfg, k):
-            setattr(engine_cfg, k, v)
+    # engine
+    if session_len is None:
+        session_len = model.session_len
 
-    tm_model = tm.TurboMind.from_pretrained(
-        model_path,
+    engine_cfg = TurbomindEngineConfig(
+        max_batch_size=max_batch_size,
```
Comment (lvhan028) [inline lmdeploy/turbomind/chat.py:85]: max_batch_size=1

File: lmdeploy/turbomind/chat.py [hunk 5]
```diff
@@ -20,23 +22,26 @@ def input_prompt(model_name):
...
+         quant_policy: int = 0,
+         cache_max_entry_count: float = 0.8,
+         cache_block_seq_len: int = 64,
+         rope_scaling_factor: float = 0.0,
+         enable_prefix_caching: bool = False,
+         session_len: int = None,
          stream_output: bool = True,
          request_output_len: int = 1024,
+         chat_template: str = None,
          chat_template_cfg: ChatTemplateConfig = None,
```
Comment (lvhan028) [inline lmdeploy/turbomind/chat.py:45]: may remove `chat_template_cfg`

File: lmdeploy/turbomind/chat.py [hunk 6]
```diff
@@ -20,23 +22,26 @@ def input_prompt(model_name):
...
+         model_format: str = None,
+         quant_policy: int = 0,
+         cache_max_entry_count: float = 0.8,
+         cache_block_seq_len: int = 64,
+         rope_scaling_factor: float = 0.0,
+         enable_prefix_caching: bool = False,
+         session_len: int = None,
          stream_output: bool = True,
          request_output_len: int = 1024,
+         chat_template: str = None,
```
Comment (lvhan028) [inline lmdeploy/turbomind/chat.py:44]: Do we need both `chat_template` and `chat_template_cfg`?

File: lmdeploy/turbomind/chat.py [hunk 7]
```diff
@@ -46,100 +51,126 @@ def main(model_path: str,
...
-        for k, v in kwargs.items():
-            if hasattr(chat_template_cfg, k):
-                setattr(chat_template_cfg, k, v)
-            else:
-                new_kwargs[k] = v
-        kwargs = new_kwargs
+        chat_template_cfg = ChatTemplateConfig(chat_template_name)
+    elif chat_template_cfg.model_name is None:
+        chat_template_cfg.model_name = chat_template_name
+    elif chat_template.capability is None:
```
Comment (lvhan028) [inline lmdeploy/turbomind/chat.py:87]: `chat_template` is str. We can not access `chat_template.capability`

File: lmdeploy/turbomind/chat.py (hunk 6 above)
Comment (irexyc) [reply lmdeploy/turbomind/chat.py:44]: `lmdeploy chat /path/to/model` will pass  `chat_template_config` 

`lmdeploy chat turbomind /path/to/model` may pass `chat_template`

Comment (lvhan028) [reply lmdeploy/turbomind/chat.py:44]: Are they the same? If so, we can change one of them

**Comment by AllentDan** [comment]
May resolve conflicts

File: lmdeploy/cli/chat.py [hunk 8]
```diff
@@ -124,16 +120,6 @@ def turbomind(args):
             ' future. Please use `lmdeploy chat` instead.')
 
         kwargs = convert_args(args)
-        from lmdeploy.model import ChatTemplateConfig
```
Comment (RunningLeon) [inline lmdeploy/cli/chat.py:127]: `lmdeploy chat turbomind  /share/140/InternLM/internlm2-chat-7b`
 each conversion will print INFO about `Register stream `, which is a bit annoying

---
double enter to end input >>> hello

<|im_start|>system
You are an AI assistant whose name is InternLM (书生·浦语).
- InternLM (书生·浦语) is a conversational language model that is developed by Shanghai AI Laboratory (上海人工智能实验室). It is designed to be helpful, honest, and harmless.
- InternLM (书生·浦语) can understand and communicate fluently in the language chosen by the user such as English and 中文.
<|im_end|>
<|im_start|>user
hello<|im_end|>
<|im_start|>assistant
**2024-06-25 15:13:17,625 - lmdeploy - INFO - Register stream callback for 1**
Hello! How can I assist you today?**2024-06-25 15:13:17,757 - lmdeploy - INFO - UN-register stream callback for 1**
double enter to end input >>> who are you
<|im_start|>user
who are you<|im_end|>
<|im_start|>assistant
**2024-06-25 15:14:08,180 - lmdeploy - INFO - Register stream callback for 1**
I am an AI language model called InternLM (书生·浦语), developed by Shanghai AI Laboratory. My purpose is to assist and communicate with you in a helpful, honest, and harmless manner. I can understand and communicate in both English and Chinese. How can I assist you today?**2024-06-25 15:14:08,880 - lmdeploy - INFO - UN-register stream callback for 1**

Comment (irexyc) [reply lmdeploy/cli/chat.py:127]: shoud be fixed in [b9a91ec](https://github.com/InternLM/lmdeploy/pull/1161/commits/b9a91ec09637ff240e5f00e8b4da829ce2b5e909)

**Review by RunningLeon — APPROVED** [review]
LGTM

**Review by lvhan028 — APPROVED** [review]

**Pull request merged**